File: services/api/src/adapters/inbound/grpc/welcome.rs
//! gRPC Welcome config — delegue au WelcomeConfigRepository.
//! Plus de SQL direct : le repo centralise la query + defaults.

use std::sync::Arc;

use tonic::{Request, Response, Status};

use sentinel_proto::welcome::v1 as proto;
use sentinel_proto::welcome::v1::welcome_service_server::WelcomeService;

use crate::ports::outbound::WelcomeConfigRepository;

pub struct WelcomeGrpc {
    pub repo: Arc<dyn WelcomeConfigRepository>,
}

#[tonic::async_trait]
impl WelcomeService for WelcomeGrpc {
    async fn get_config(
        &self,
        request: Request<proto::GetConfigRequest>,
    ) -> Result<Response<proto::WelcomeConfig>, Status> {
        let cfg = self.repo.get_config(&request.into_inner().guild_id).await
            .map_err(|e| Status::internal(format!("get welcome config: {e}")))?;

        Ok(Response::new(proto::WelcomeConfig {
            guild_id: cfg.guild_id,
            welcome_enabled: cfg.welcome_enabled,
            welcome_channel_id: cfg.welcome_channel_id,
            welcome_message: cfg.welcome_message,
            welcome_embed_color: cfg.welcome_embed_color,
            welcome_dm_enabled: cfg.welcome_dm_enabled,
            welcome_dm_message: cfg.welcome_dm_message,
            leave_enabled: cfg.leave_enabled,
            leave_channel_id: cfg.leave_channel_id,
            leave_message: cfg.leave_message,
            rules_enabled: cfg.rules_enabled,
            rules_channel_id: cfg.rules_channel_id,
            rules_message: cfg.rules_message,
            rules_role_id: cfg.rules_role_id,
            rules_button_label: cfg.rules_button_label,
            counter_enabled: cfg.counter_enabled,
            counter_channel_id: cfg.counter_channel_id,
            counter_format: cfg.counter_format,
            anniversary_enabled: cfg.anniversary_enabled,
            anniversary_channel_id: cfg.anniversary_channel_id,
            anniversary_message: cfg.anniversary_message,
            rejoin_message: cfg.rejoin_message,
        }))
    }
}

File: apps/web/src/types/welcome.ts
/**
 * Configuration Welcome / Onboarding (mappee 1:1 sur l API
 * services/api/src/adapters/inbound/http/handlers/welcome.rs).
 *
 * 5 sections :
 * - welcome   : message de bienvenue dans un salon (embed + DM)
 * - leave     : message de depart
 * - rules     : verification gate (lecture des regles + role attribue)
 * - counter   : compteur de membres dans un nom de salon
 * - anniversary: anniversaire d arrivee sur le serveur
 */
export interface WelcomeConfig {
  guild_id: string;

  welcome_enabled: boolean;
  welcome_channel_id: string | null;
  welcome_message: string;
  welcome_embed_color: string;
  welcome_dm_enabled: boolean;
  welcome_dm_message: string;
  welcome_title: string;
  welcome_image_url: string;
  welcome_footer_text: string;

  leave_enabled: boolean;
  leave_channel_id: string | null;
  leave_message: string;
  leave_title: string;
  leave_image_url: string;
  leave_footer_text: string;

  rules_enabled: boolean;
  rules_channel_id: string | null;
  rules_message: string;
  rules_role_id: string | null;
  rules_button_label: string;

  counter_enabled: boolean;
  counter_channel_id: string | null;
  counter_format: string;

  anniversary_enabled: boolean;
  anniversary_channel_id: string | null;
  anniversary_message: string;
  anniversary_title: string;
  anniversary_image_url: string;
  anniversary_footer_text: string;

  rejoin_message: string;
  rejoin_title: string;
  rejoin_image_url: string;
  rejoin_footer_text: string;
}

/** Payload PATCH-like : tous les champs sont optionnels (merge cote API). */
export type SaveWelcomeConfigParams = Partial<Omit<WelcomeConfig, "guild_id">>;
